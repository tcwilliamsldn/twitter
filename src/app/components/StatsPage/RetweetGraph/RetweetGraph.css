.graph {
    max-width: 1080px;
    background-color: #ffffff;
    margin: 12px;
    box-shadow: 0px 2px 10px #9a9a9a;
}

.infobar {
    display: flex;
    padding: 24px 48px;
    justify-content: space-between;
    align-items: center;

    & > h1 {
        font-weight: bold;
        font-size: 18px;
        padding: 0;
        margin: 0;
    }

    & > span {
        color: #737373;
    }
}

.loading {
    padding: 48px 0 96px;
    font-size: 24px;
    text-align: center;
    color: #757575;
    animation: fadeInOut 1s infinite;
}

@keyframes fadeInOut {
    0% {
        opacity: 1;
    }
    50% {
        opacity: 0.1;
    }
    100% {
        opacity: 1;
    }
}
