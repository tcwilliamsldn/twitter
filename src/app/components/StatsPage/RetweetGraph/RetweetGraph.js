import React, { PureComponent } from 'react';
import styles from './RetweetGraph.css';
import Tooltip from '../Tooltip/Tooltip';

// Semiotic and D3 make up well over half the bundle, and as the library is transpiled
// webpack couldn't tree shake the code. Only importing the file used and using
// dynamic imports reduced the main bundle from 947 kB to 328 kB (89 kB gzipped).

class RetweetGraph extends PureComponent {
    state = {};

    componentDidMount() {
        import(/* webpackChunkName: "xyFrame" */ 'semiotic/lib/ResponsiveXYFrame').then(
            ResponsiveXYFrame =>
                this.setState({ ResponsiveXYFrame: ResponsiveXYFrame.default })
        );
    }

    render() {
        const { ResponsiveXYFrame } = this.state;

        const { tweets } = this.props;
        const totalRetweets = tweets.reduce(
            (total, current) => total + current.retweet_count,
            0
        );

        const coordinates = tweets.map(({ created_at, retweet_count }) => ({
            date: Date.parse(created_at),
            retweet_count
        }));

        return (
            <div className={styles.graph}>
                <div className={styles.infobar}>
                    <h1>Retweets for your last 50 tweets</h1>
                    <span>{`${totalRetweets} retweets`}</span>
                </div>
                {ResponsiveXYFrame ? (
                    <ResponsiveXYFrame
                        size={[700, 300]}
                        responsiveWidth
                        areas={[{ coordinates }]}
                        xAccessor="date"
                        yAccessor="retweet_count"
                        areaStyle={{
                            fill: '#e0e9ff',
                            stroke: '#306aff',
                            strokeWidth: '2px'
                        }}
                        hoverAnnotation={[
                            { type: 'frame-hover', color: 'transparent' }
                        ]}
                        tooltipContent={props => <Tooltip {...props} />}
                    />
                ) : (
                    <div className={styles.loading}>Loading Graph...</div>
                )}
            </div>
        );
    }
}

export default RetweetGraph;
