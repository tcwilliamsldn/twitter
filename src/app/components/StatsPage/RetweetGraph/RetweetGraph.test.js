import RetweetGraph from './RetweetGraph';
import React from 'react';
import { shallow } from 'enzyme';

describe('RetweetGraph', function() {
    it('renders a loading state', function() {
        const tweets = [
            {
                created_at: 'Thu Jun 18 19:44:33 +0000 2009',
                retweet_count: 100
            },
            { created_at: 'Thu Jun 18 19:44:33 +0000 2009', retweet_count: 10 }
        ];

        const component = shallow(<RetweetGraph tweets={tweets} />);
        expect(component).toMatchSnapshot();
    });

    it('renders a retweet graph once library loaded', function() {
        const tweets = [
            {
                created_at: 'Thu Jun 18 19:44:33 +0000 2009',
                retweet_count: 100
            },
            { created_at: 'Thu Jun 18 19:44:33 +0000 2009', retweet_count: 10 }
        ];

        const ResponsiveXYFrame = jest.fn();
        const component = shallow(<RetweetGraph tweets={tweets} />);
        component.setState({ ResponsiveXYFrame });
        expect(component).toMatchSnapshot();
    });
});
