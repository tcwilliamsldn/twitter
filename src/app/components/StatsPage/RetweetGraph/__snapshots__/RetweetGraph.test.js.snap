// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RetweetGraph renders a loading state 1`] = `
<div
  className="graph"
>
  <div
    className="infobar"
  >
    <h1>
      Retweets for your last 50 tweets
    </h1>
    <span>
      110 retweets
    </span>
  </div>
  <div
    className="loading"
  >
    Loading Graph...
  </div>
</div>
`;

exports[`RetweetGraph renders a retweet graph once library loaded 1`] = `
<div
  className="graph"
>
  <div
    className="infobar"
  >
    <h1>
      Retweets for your last 50 tweets
    </h1>
    <span>
      110 retweets
    </span>
  </div>
  <mockConstructor
    areaStyle={
      Object {
        "fill": "#e0e9ff",
        "stroke": "#306aff",
        "strokeWidth": "2px",
      }
    }
    areas={
      Array [
        Object {
          "coordinates": Array [
            Object {
              "date": 1245354273000,
              "retweet_count": 100,
            },
            Object {
              "date": 1245354273000,
              "retweet_count": 10,
            },
          ],
        },
      ]
    }
    hoverAnnotation={
      Array [
        Object {
          "color": "transparent",
          "type": "frame-hover",
        },
      ]
    }
    responsiveWidth={true}
    size={
      Array [
        700,
        300,
      ]
    }
    tooltipContent={[Function]}
    xAccessor="date"
    yAccessor="retweet_count"
  />
</div>
`;
